File: pkg/usecase/todo.go
package usecase

import (
	"context"
	"sort"
	"sync"

	"github.com/gogo/status"
	"google.golang.org/grpc/codes"

	"github.com/iwaltgen/grpc-go-web-todo/pkg/entity"
	"github.com/iwaltgen/grpc-go-web-todo/pkg/event"
	"github.com/iwaltgen/grpc-go-web-todo/pkg/log"
	"github.com/iwaltgen/grpc-go-web-todo/pkg/repository"
)

// Todo is domain logic for todo entity
type Todo struct {
	logger  *log.Logger
	repo    *repository.Todo
	emitter *event.Emitter
}

var (
	defaultTodo     *Todo
	defaultTodoOnce sync.Once
)

// DefaultTodo is get default todo usecase
func DefaultTodo() *Todo {
	defaultTodoOnce.Do(func() {
		defaultTodo = createTodo()
	})
	return defaultTodo
}

func newTodo(repo *repository.Todo) *Todo {
	return &Todo{
		logger:  log.L("usecase.todo"),
		repo:    repo,
		emitter: &event.Emitter{},
	}
}

// ListTodos is get all todo entity
func (s *Todo) ListTodos(ctx context.Context) (ret []*entity.Todo, err error) {
	ret, err = s.repo.FindAll(ctx)
	sort.Sort(sort.Reverse(todoSortByCreatedAt(ret)))
	return ret, err
}

// CreateTodo is create todo entity
func (s *Todo) CreateTodo(ctx context.Context, v *entity.Todo) error {
	err := s.repo.Create(ctx, v)
	if err == nil {
		s.emitter.Publish(event.EventCreate, v)
	}
	return err
}

// UpdateTodo is update todo entity
func (s *Todo) UpdateTodo(ctx context.Context, v *entity.Todo) error {
	err := s.repo.Update(ctx, v)
	if err == nil {
		s.emitter.Publish(event.EventUpdate, v)
	}
	return err
}

// DeleteTodo is delete todo entity
func (s *Todo) DeleteTodo(ctx context.Context, id string) error {
	if ret, ok := s.repo.FindByID(ctx, id); ok {
		err := s.repo.DeleteByID(ctx, id)
		if err == nil {
			s.emitter.Publish(event.EventDelete, ret)
		}
		return err
	}
	return status.Errorf(codes.NotFound, "not found error: %s", id)
}

// Subscribe is subscribe entity change event
func (s *Todo) Subscribe(handler event.Subscription, evt event.Event) func() {
	return s.emitter.Subscribe(handler, event.WithEvent(evt))
}

// todo entity sort by modified time
type todoSortByCreatedAt []*entity.Todo

func (s todoSortByCreatedAt) Len() int {
	return len(s)
}

func (s todoSortByCreatedAt) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

func (s todoSortByCreatedAt) Less(i, j int) bool {
	return s[i].CreatedAt.Before(s[j].CreatedAt)
}

File: src/usecase/todo.ts
import { writable, derived, Readable } from 'svelte/store';
import { todoClient } from './grpc';
import type { Error, Status } from '../grpc';
import type { Todo } from '../entity';
import type { EventTodo } from './value';
import { now } from './timestamp';
import { Event } from '../api/todo/v1';

export class TodoService {
  private todos = writable([] as Todo[]);
  private client = todoClient;

  constructor() {
    this.subscribe();
  }

  async list(): Promise<Todo[]> {
    const items = await this.client.list();
    this.todos.set(items);
    return items;
  }

  async create(todo: Todo): Promise<void> {
    const nowTs = now();
    todo.modifiedAt = nowTs;
    todo.createdAt = nowTs;
    await this.client.create(todo);
  }

  async update(todo: Todo): Promise<void> {
    todo.modifiedAt = now();
    await this.client.update(todo);
  }

  async delete(todo: Todo): Promise<void> {
    await this.client.delete(todo);
  }

  store(): Readable<Todo[]> {
    return derived(this.todos, ($todos) => $todos);
  }

  private subscribe() {
    this.client.subscribe().subscribe(
      (evt: EventTodo) => {
        this.todos.update((list) => {
          switch (evt.event) {
            case Event.EVENT_CREATE:
              return [evt.todo, ...list];

            case Event.EVENT_UPDATE:
              return list.map((item) => {
                if (item.id === evt.todo.id) {
                  return evt.todo;
                }
                return item;
              });

            case Event.EVENT_DELETE:
              return list.filter((item) => item.id !== evt.todo.id);
          }
          return list.concat(evt.todo);
        });
      },
      (error: Error | Status) => {
        console.log('todo service subscribe error: ', error);
      },
    );
  }
}

export const todoService = new TodoService();
